File: sdk/javascript/gmaps-scraper-sdk.js
/**
 * Google Maps Lead Scraper — JavaScript / Node.js SDK
 * =====================================================
 * Works in Node.js 18+ (native fetch) and browser.
 * No dependencies — uses built-in fetch + EventSource.
 *
 * Install (Node.js): just copy this file, no npm needed.
 *
 * USAGE EXAMPLES:
 *
 * const client = new LeadScraperClient("http://localhost:8000");
 *
 * // 1. Simple sync scrape
 * const leads = await client.scrape("Dentists in Mathura", { maxResults: 50 });
 *
 * // 2. Async job with polling
 * const job = await client.createJob("Gyms in Delhi", { maxResults: 200 });
 * const results = await client.waitForJob(job.job_id);
 *
 * // 3. Live stream — get leads as they're scraped
 * await client.streamJob("Restaurants in Agra", { maxResults: 100 }, (lead) => {
 *   console.log("Found:", lead.name, lead.phone);
 * });
 */

class LeadScraperError extends Error {
  constructor(message, statusCode = null) {
    super(message);
    this.name = "LeadScraperError";
    this.statusCode = statusCode;
  }
}

class LeadScraperClient {
  /**
   * @param {string} baseUrl  - Your API URL e.g. "http://localhost:8000"
   * @param {object} options
   * @param {number} options.timeout    - Fetch timeout ms (default 600000 = 10min)
   * @param {string} options.apiKey     - Optional API key
   */
  constructor(baseUrl = "http://localhost:8000", options = {}) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.timeout = options.timeout ?? 600_000;
    this.apiKey = options.apiKey ?? null;
  }

  // ── Core API ───────────────────────────────────────────────────────────────

  /**
   * Synchronous scrape — waits for all results.
   * @param {string} query        - e.g. "Dentists in Mathura"
   * @param {object} opts
   * @param {number} opts.maxResults     - Max leads (default 50, max 1000)
   * @param {boolean} opts.extractDetails - Full details per place (default true)
   * @returns {Promise<Array>} Array of lead objects
   *
   * @example
   * const leads = await client.scrape("Dentists in Mathura", { maxResults: 50 });
   * leads.forEach(l => console.log(l.name, l.phone, l.rating));
   */
  async scrape(query, opts = {}) {
    const data = await this._post("/scrape", {
      query,
      max_results: opts.maxResults ?? 50,
      extract_details: opts.extractDetails ?? true,
    });
    return data.data ?? [];
  }

  /**
   * Submit an async job — returns immediately with job info.
   * @param {string} query
   * @param {object} opts
   * @param {number}  opts.maxResults
   * @param {boolean} opts.extractDetails
   * @param {string}  opts.webhookUrl - Optional webhook to POST results to when done
   * @returns {Promise<object>} { job_id, status, query, created_at }
   *
   * @example
   * const job = await client.createJob("Gyms in Delhi", { maxResults: 200 });
   * console.log("Job ID:", job.job_id);
   */
  async createJob(query, opts = {}) {
    return this._post("/jobs", {
      query,
      max_results: opts.maxResults ?? 100,
      extract_details: opts.extractDetails ?? true,
      webhook_url: opts.webhookUrl ?? null,
    });
  }

  /**
   * Poll a job's current status.
   * @returns {Promise<object>} { job_id, status, progress, results_so_far, ... }
   */
  async pollJob(jobId) {
    return this._get(`/jobs/${jobId}`);
  }

  /**
   * Get full results for a completed job.
   * @returns {Promise<Array>} Array of leads
   */
  async getResults(jobId) {
    const data = await this._get(`/jobs/${jobId}/results`);
    return data.data ?? [];
  }

  /**
   * Submit job and poll until complete. Logs progress to console.
   * @param {string} jobId
   * @param {object} opts
   * @param {number}  opts.pollInterval - ms between polls (default 3000)
   * @param {boolean} opts.verbose      - log progress (default true)
   * @returns {Promise<Array>} Final leads array
   *
   * @example
   * const job = await client.createJob("Gyms in Delhi", { maxResults: 200 });
   * const leads = await client.waitForJob(job.job_id);
   */
  async waitForJob(jobId, opts = {}) {
    const interval = opts.pollInterval ?? 3000;
    const verbose = opts.verbose ?? true;

    while (true) {
      const status = await this.pollJob(jobId);
      const { status: st, progress, results_so_far, execution_time } = status;

      if (verbose) {
        process.stdout && process.stdout.write
          ? process.stdout.write(`\r  [${st.toUpperCase()}] ${progress}% — ${results_so_far} leads...`)
          : console.log(`[${st}] ${progress}% — ${results_so_far} leads`);
      }

      if (st === "completed") {
        if (verbose) console.log(`\n  ✅ Done! ${results_so_far} leads in ${execution_time}`);
        return this.getResults(jobId);
      }

      if (st === "failed") {
        throw new LeadScraperError(`Job failed: ${status.error}`);
      }

      await new Promise(r => setTimeout(r, interval));
    }
  }

  /**
   * Submit a job and stream results live via SSE.
   * Callback fires for EVERY lead as it's found.
   *
   * @param {string}   query
   * @param {object}   opts         - same as createJob opts
   * @param {function} onLead       - called with each lead dict
   * @param {function} onDone       - called when complete with { total }
   * @returns {Promise<void>}
   *
   * @example
   * await client.streamJob("Restaurants in Agra", { maxResults: 100 },
   *   (lead) => console.log("Got:", lead.name, lead.phone),
   *   (summary) => console.log("Total:", summary.total)
   * );
   */
  async streamJob(query, opts = {}, onLead = null, onDone = null) {
    const job = await this.createJob(query, opts);
    const jobId = job.job_id;

    return new Promise((resolve, reject) => {
      const url = `${this.baseUrl}/jobs/${jobId}/stream`;

      // Use EventSource if available (browser), else fetch stream (Node)
      if (typeof EventSource !== "undefined") {
        const es = new EventSource(url);
        es.onmessage = (e) => {
          try {
            const event = JSON.parse(e.data);
            if (event.type === "result" && onLead) onLead(event.lead);
            if (event.type === "done") {
              es.close();
              if (onDone) onDone({ total: event.total, status: event.status });
              resolve();
            }
            if (event.type === "error") {
              es.close();
              reject(new LeadScraperError(event.error));
            }
          } catch (err) { /* skip malformed */ }
        };
        es.onerror = (e) => { es.close(); reject(new LeadScraperError("SSE connection error")); };
      } else {
        // Node.js fetch streaming
        fetch(url, { signal: AbortSignal.timeout(this.timeout) })
          .then(async (res) => {
            const reader = res.body.getReader();
            const decoder = new TextDecoder();
            let buffer = "";
            while (true) {
              const { done, value } = await reader.read();
              if (done) break;
              buffer += decoder.decode(value, { stream: true });
              const lines = buffer.split("\n");
              buffer = lines.pop();
              for (const line of lines) {
                if (!line.startsWith("data:")) continue;
                try {
                  const event = JSON.parse(line.slice(5).trim());
                  if (event.type === "result" && onLead) onLead(event.lead);
                  if (event.type === "done") {
                    if (onDone) onDone({ total: event.total, status: event.status });
                    resolve(); return;
                  }
                  if (event.type === "error") {
                    reject(new LeadScraperError(event.error)); return;
                  }
                } catch { /* skip */ }
              }
            }
            resolve();
          })
          .catch(reject);
      }
    });
  }

  /**
   * List all jobs on the server.
   * @returns {Promise<Array>}
   */
  async listJobs() {
    return this._get("/jobs");
  }

  /**
   * Delete a job.
   */
  async deleteJob(jobId) {
    const res = await fetch(`${this.baseUrl}/jobs/${jobId}`, {
      method: "DELETE",
      headers: this._headers(),
    });
    if (!res.ok) throw new LeadScraperError(`Delete failed: ${res.status}`);
    return res.json();
  }

  /**
   * Get download URL for CSV export.
   * In Node.js, this downloads and saves the file.
   * In browser, triggers a file download.
   */
  async downloadCsv(jobId, filename = null) {
    const url = `${this.baseUrl}/jobs/${jobId}/export?format=csv`;
    return this._download(url, filename ?? `leads_${jobId.slice(0,8)}.csv`);
  }

  /**
   * Get download URL for Excel export.
   */
  async downloadExcel(jobId, filename = null) {
    const url = `${this.baseUrl}/jobs/${jobId}/export?format=xlsx`;
    return this._download(url, filename ?? `leads_${jobId.slice(0,8)}.xlsx`);
  }

  /**
   * Check API health.
   */
  async health() {
    return this._get("/health");
  }

  /**
   * Get scraping metrics.
   */
  async metrics() {
    return this._get("/metrics");
  }

  // ── Utility ────────────────────────────────────────────────────────────────

  /** Convert leads array to CSV string */
  leadsToCSV(leads) {
    if (!leads.length) return "";
    const fields = Object.keys(leads[0]);
    const escape = (v) => {
      if (v === null || v === undefined) return "";
      const s = Array.isArray(v) ? v.join(" | ") : String(v);
      return s.includes(",") || s.includes('"') || s.includes("\n")
        ? `"${s.replace(/"/g, '""')}"` : s;
    };
    return [
      fields.join(","),
      ...leads.map(l => fields.map(f => escape(l[f])).join(","))
    ].join("\n");
  }

  // ── Internal ───────────────────────────────────────────────────────────────

  _headers() {
    const h = { "Content-Type": "application/json", "Accept": "application/json" };
    if (this.apiKey) h["X-API-Key"] = this.apiKey;
    return h;
  }

  async _post(path, body) {
    let res;
    try {
      res = await fetch(`${this.baseUrl}${path}`, {
        method: "POST",
        headers: this._headers(),
        body: JSON.stringify(body),
        signal: AbortSignal.timeout(this.timeout),
      });
    } catch (e) {
      throw new LeadScraperError(`Network error: ${e.message}`);
    }
    if (!res.ok) {
      const err = await res.json().catch(() => ({}));
      throw new LeadScraperError(err.detail ?? `HTTP ${res.status}`, res.status);
    }
    return res.json();
  }

  async _get(path) {
    let res;
    try {
      res = await fetch(`${this.baseUrl}${path}`, {
        headers: this._headers(),
        signal: AbortSignal.timeout(this.timeout),
      });
    } catch (e) {
      throw new LeadScraperError(`Network error: ${e.message}`);
    }
    if (!res.ok) {
      const err = await res.json().catch(() => ({}));
      throw new LeadScraperError(err.detail ?? `HTTP ${res.status}`, res.status);
    }
    return res.json();
  }

  async _download(url, filename) {
    // Browser
    if (typeof window !== "undefined") {
      const a = document.createElement("a");
      a.href = url; a.download = filename;
      document.body.appendChild(a); a.click();
      document.body.removeChild(a);
      return url;
    }
    // Node.js
    const { writeFileSync } = await import("fs");
    const res = await fetch(url, { headers: this._headers() });
    if (!res.ok) throw new LeadScraperError(`Download failed: ${res.status}`);
    const buf = await res.arrayBuffer();
    writeFileSync(filename, Buffer.from(buf));
    console.log(`✅ Saved to ${filename}`);
    return filename;
  }
}

// ── Quick Node.js CLI demo ─────────────────────────────────────────────────────

if (typeof process !== "undefined" && process.argv?.[1]?.endsWith("gmaps-scraper-sdk.js")) {
  const baseUrl = process.argv[2] ?? "http://localhost:8000";
  const query   = process.argv[3] ?? "Dentists in Mathura";
  const count   = parseInt(process.argv[4] ?? "20");

  (async () => {
    console.log(`\n🔍 Scraping: "${query}" (max ${count})`);
    console.log(`🌐 API: ${baseUrl}\n`);

    const client = new LeadScraperClient(baseUrl);

    const h = await client.health();
    console.log(`✅ API healthy | Browsers: ${h.browser_pool.active}/${h.browser_pool.pool_size}\n`);

    console.log("🚀 Streaming results live...\n");
    const all = [];
    await client.streamJob(query, { maxResults: count },
      (lead) => {
        all.push(lead);
        console.log(`  [${all.length}] ${lead.name} | ⭐${lead.rating ?? "?"} | 📞${lead.phone ?? "N/A"}`);
      },
      (summary) => console.log(`\n🏁 Done! Total: ${summary.total}`)
    );

    const csv = client.leadsToCSV(all);
    const { writeFileSync } = await import("fs");
    const fname = `leads_${query.slice(0,15).replace(/ /g,"_")}.csv`;
    writeFileSync(fname, csv);
    console.log(`\n💾 Saved: ${fname}`);
  })().catch(console.error);
}

// Export for Node.js
if (typeof module !== "undefined") module.exports = { LeadScraperClient, LeadScraperError };

File: sdk/python/gmaps_scraper_sdk.py
"""
Google Maps Lead Scraper — Python SDK
======================================
Zero dependencies beyond `requests` (sync) or `httpx` (async).
Install: pip install requests httpx sseclient-py

Usage — sync:
    from gmaps_scraper_sdk import LeadScraperClient
    client = LeadScraperClient("http://localhost:8000")
    leads  = client.scrape("Dentists in Mathura", max_results=50)
    print(leads[0])

Usage — async job:
    job = client.create_job("Gyms in Delhi", max_results=200)
    results = client.wait_for_job(job["job_id"])

Usage — SSE live stream:
    for lead in client.stream_job("Restaurants in Agra", max_results=100):
        print(lead["name"], lead["phone"])
"""

import time
import json
import csv
import io
from typing import List, Dict, Optional, Iterator, Generator

try:
    import requests
except ImportError:
    raise ImportError("Run: pip install requests")


class LeadScraperError(Exception):
    pass


class LeadScraperClient:
    """
    Synchronous Python client for the Google Maps Lead Scraper API.

    Args:
        base_url: Your deployed API URL. e.g. "http://localhost:8000"
                  or "https://your-app.justrunmyapp.com"
        timeout:  Request timeout in seconds (default 600 for large scrapes)
        api_key:  Optional API key header (set on server via middleware if needed)
    """

    def __init__(
        self,
        base_url: str = "http://localhost:8000",
        timeout: int = 600,
        api_key: Optional[str] = None,
    ):
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self.session = requests.Session()
        self.session.headers.update({
            "Content-Type": "application/json",
            "Accept": "application/json",
        })
        if api_key:
            self.session.headers["X-API-Key"] = api_key

    # ── Core methods ──────────────────────────────────────────────────────────

    def scrape(
        self,
        query: str,
        max_results: int = 50,
        extract_details: bool = True,
    ) -> List[Dict]:
        """
        Synchronous scrape — blocks until all results are ready.

        Args:
            query:           e.g. "Dentists in Mathura"
            max_results:     How many leads to return (max 1000)
            extract_details: True = full data per place (slower but complete)
                             False = name/rating/url only (very fast)

        Returns:
            List of lead dicts with all fields.

        Example:
            client = LeadScraperClient("http://localhost:8000")
            leads = client.scrape("Dentists in Mathura", max_results=50)
            for lead in leads:
                print(f"{lead['name']} | {lead['phone']} | {lead['rating']}")
        """
        resp = self._post("/scrape", {
            "query": query,
            "max_results": max_results,
            "extract_details": extract_details,
        })
        return resp.get("data", [])

    def create_job(
        self,
        query: str,
        max_results: int = 100,
        extract_details: bool = True,
        webhook_url: Optional[str] = None,
    ) -> Dict:
        """
        Submit an async job. Returns immediately with a job_id.
        Use poll_job() or wait_for_job() to get results.

        Returns dict with keys: job_id, status, query, created_at
        """
        return self._post("/jobs", {
            "query": query,
            "max_results": max_results,
            "extract_details": extract_details,
            "webhook_url": webhook_url,
        })

    def poll_job(self, job_id: str) -> Dict:
        """Get current job status and progress (0-100%)."""
        return self._get(f"/jobs/{job_id}")

    def get_results(self, job_id: str) -> List[Dict]:
        """
        Get full results for a completed job.
        Raises LeadScraperError if job not yet done.
        """
        data = self._get(f"/jobs/{job_id}/results")
        return data.get("data", [])

    def wait_for_job(
        self,
        job_id: str,
        poll_interval: float = 3.0,
        verbose: bool = True,
    ) -> List[Dict]:
        """
        Poll until job is complete, then return results.

        Args:
            job_id:        From create_job()
            poll_interval: Seconds between status checks
            verbose:       Print progress to console

        Example:
            job = client.create_job("Gyms in Delhi", max_results=200)
            leads = client.wait_for_job(job["job_id"], verbose=True)
        """
        while True:
            status = self.poll_job(job_id)
            pct = status.get("progress", 0)
            st = status.get("status", "")
            count = status.get("results_so_far", 0)

            if verbose:
                print(f"  [{st.upper()}] {pct}% — {count} leads so far...")

            if st == "completed":
                if verbose:
                    print(f"  ✅ Done! {count} leads scraped in {status.get('execution_time')}")
                return self.get_results(job_id)

            if st == "failed":
                raise LeadScraperError(f"Job failed: {status.get('error')}")

            time.sleep(poll_interval)

    def stream_job(
        self,
        query: str,
        max_results: int = 100,
        extract_details: bool = True,
    ) -> Generator[Dict, None, None]:
        """
        Submit a job and yield each lead AS IT'S SCRAPED via SSE.
        No waiting — results come in real-time.

        Example:
            for lead in client.stream_job("Restaurants in Agra", max_results=100):
                print(f"Found: {lead['name']} — {lead['phone']}")
        """
        job = self.create_job(query, max_results, extract_details)
        job_id = job["job_id"]

        url = f"{self.base_url}/jobs/{job_id}/stream"
        with self.session.get(url, stream=True, timeout=self.timeout) as resp:
            resp.raise_for_status()
            for line in resp.iter_lines():
                if not line:
                    continue
                line = line.decode("utf-8") if isinstance(line, bytes) else line
                if line.startswith("data:"):
                    try:
                        event = json.loads(line[5:].strip())
                        etype = event.get("type")
                        if etype == "result":
                            yield event["lead"]
                        elif etype == "done":
                            break
                        elif etype == "error":
                            raise LeadScraperError(event.get("error", "Unknown error"))
                    except json.JSONDecodeError:
                        continue

    def list_jobs(self) -> List[Dict]:
        """List all jobs (queued, running, completed, failed)."""
        return self._get("/jobs")

    def delete_job(self, job_id: str) -> Dict:
        """Delete a job and its results."""
        resp = self.session.delete(f"{self.base_url}/jobs/{job_id}", timeout=30)
        resp.raise_for_status()
        return resp.json()

    def download_csv(self, job_id: str, filepath: str) -> str:
        """
        Download results as CSV file.

        Example:
            client.download_csv("abc-123", "dentists_mathura.csv")
        """
        resp = self.session.get(f"{self.base_url}/jobs/{job_id}/export?format=csv", timeout=60)
        resp.raise_for_status()
        with open(filepath, "wb") as f:
            f.write(resp.content)
        print(f"✅ Saved to {filepath}")
        return filepath

    def download_excel(self, job_id: str, filepath: str) -> str:
        """
        Download results as Excel (.xlsx) file.

        Example:
            client.download_excel("abc-123", "dentists_mathura.xlsx")
        """
        resp = self.session.get(f"{self.base_url}/jobs/{job_id}/export?format=xlsx", timeout=60)
        resp.raise_for_status()
        with open(filepath, "wb") as f:
            f.write(resp.content)
        print(f"✅ Saved to {filepath}")
        return filepath

    def health(self) -> Dict:
        """Check API health and browser pool status."""
        return self._get("/health")

    def metrics(self) -> Dict:
        """Get scraping statistics."""
        return self._get("/metrics")

    # ── Utilities ─────────────────────────────────────────────────────────────

    def leads_to_csv_string(self, leads: List[Dict]) -> str:
        """Convert a list of lead dicts to a CSV string (no file needed)."""
        if not leads:
            return ""
        fields = list(leads[0].keys())
        buf = io.StringIO()
        writer = csv.DictWriter(buf, fieldnames=fields, extrasaction="ignore")
        writer.writeheader()
        writer.writerows(leads)
        return buf.getvalue()

    # ── Internal ──────────────────────────────────────────────────────────────

    def _post(self, path: str, payload: dict) -> dict:
        try:
            resp = self.session.post(
                f"{self.base_url}{path}",
                json=payload,
                timeout=self.timeout,
            )
            resp.raise_for_status()
            return resp.json()
        except requests.HTTPError as e:
            try:
                detail = e.response.json().get("detail", str(e))
            except Exception:
                detail = str(e)
            raise LeadScraperError(f"HTTP {e.response.status_code}: {detail}")
        except requests.RequestException as e:
            raise LeadScraperError(f"Request failed: {e}")

    def _get(self, path: str) -> dict:
        try:
            resp = self.session.get(
                f"{self.base_url}{path}",
                timeout=self.timeout,
            )
            resp.raise_for_status()
            return resp.json()
        except requests.HTTPError as e:
            try:
                detail = e.response.json().get("detail", str(e))
            except Exception:
                detail = str(e)
            raise LeadScraperError(f"HTTP {e.response.status_code}: {detail}")
        except requests.RequestException as e:
            raise LeadScraperError(f"Request failed: {e}")


# ── Quick CLI usage ────────────────────────────────────────────────────────────

if __name__ == "__main__":
    import sys

    base = sys.argv[1] if len(sys.argv) > 1 else "http://localhost:8000"
    query = sys.argv[2] if len(sys.argv) > 2 else "Dentists in Mathura"
    count = int(sys.argv[3]) if len(sys.argv) > 3 else 20

    print(f"\n🔍 Scraping: '{query}' (max {count} results)")
    print(f"🌐 API: {base}\n")

    client = LeadScraperClient(base)

    print("📡 Checking health...")
    h = client.health()
    print(f"   Status: {h['status']} | Browsers: {h['browser_pool']['active']}/{h['browser_pool']['pool_size']}\n")

    print("🚀 Submitting async job...")
    job = client.create_job(query, max_results=count)
    print(f"   Job ID: {job['job_id']}\n")

    print("⏳ Waiting for results...")
    leads = client.wait_for_job(job["job_id"], verbose=True)

    print(f"\n📋 First 3 results:")
    for lead in leads[:3]:
        print(f"  • {lead.get('name')} | ⭐{lead.get('rating')} | 📞{lead.get('phone')} | 🌐{lead.get('website')}")

    out = f"leads_{query[:20].replace(' ','_')}.csv"
    client.download_csv(job["job_id"], out)
